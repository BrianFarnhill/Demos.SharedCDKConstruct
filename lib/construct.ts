import * as cdk from "aws-cdk-lib";
import { aws_lambda as lambda } from "aws-cdk-lib";
import { Construct } from "constructs";
import * as path from "path";

export interface ConstructProps {
  /**
   * The visibility timeout to be configured on the SQS Queue, in seconds.
   *
   * @default Duration.seconds(3)
   */
  timeout?: cdk.Duration;

  /**
   * Whether or not a timestamp should be added to the description to force a lambda update on each deployment
   */
  includeTimestamp?: boolean;
}

export class DemoFunction extends Construct {

  LambdaFunction: lambda.Function

  constructor(scope: Construct, id: string, props: ConstructProps = {}) {
    super(scope, id);

    if (props.timeout && props.timeout.toSeconds() > 300) {
      throw new Error('This lambda must timeout in 5 minutes or less');
    }

    let description = 'An example function that comes from a shared CDK construct';
    if (props.includeTimestamp === undefined || props.includeTimestamp === true) {
      description += ` - last synthesised at ${Date.now().toString()}`;
    }

    this.LambdaFunction = new lambda.Function(this, "DemoFunction", {
      code: lambda.Code.fromAsset(path.resolve(__dirname, "./lambda")),
      handler: "index.handler",
      description,
      runtime: lambda.Runtime.NODEJS_14_X,
      timeout: props.timeout,
    });
  }
}
