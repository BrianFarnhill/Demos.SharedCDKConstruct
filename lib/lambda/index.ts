import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';

export async function handler(event: APIGatewayProxyEvent, context: Context): Promise<APIGatewayProxyResult> {
  return {
    body: 'Hello from a Lambda Function and Brian was here!',
    statusCode: 200,
  };
}
