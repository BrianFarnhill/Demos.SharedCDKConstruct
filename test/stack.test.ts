import { Template } from '@aws-cdk/assertions';
import * as cdk from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { DemoFunction } from '../lib/construct';


/**
 * A simple snapshot test
 */
 test('Matches SnapShot', () => {
  const app = new cdk.App();

  const stack = new cdk.Stack(app, "TestStack");
  new DemoFunction(stack, "TestFunction", {
    includeTimestamp: false,
  });
  const testStackOutput = app.synth().getStackArtifact("TestStack").template;

  expect(Template.fromJSON(testStackOutput)).toMatchSnapshot();
});

/**
 * A Simple fine-grained assertion
 */
test('Includes a lambda function in the output', () => {
  const app = new cdk.App();

  const stack = new cdk.Stack(app, "TestStack");
  new DemoFunction(stack, "TestFunction", {});
  const testStackOutput = app.synth().getStackArtifact("TestStack").template;
  const assert = Template.fromJSON(testStackOutput)
  assert.hasResource("AWS::Lambda::Function", {
    Properties: {
      Handler: "index.handler",
    },
  });
});

/**
 * Examples of basic validation tests
 */
test('Accepts a valid timeout value under 5 minutes', () => {
  const app = new cdk.App();
  expect(() => {
    const stack = new cdk.Stack(app, "TestStack");
    new DemoFunction(stack, "TestFunction", {
      timeout: Duration.minutes(3),
    })
  }).not.toThrow();
});
test('Rejects an invalid timeout value over 5 minutes', () => {
  const app = new cdk.App();
  expect(() => {
    const stack = new cdk.Stack(app, "TestStack");
    new DemoFunction(stack, "TestFunction", {
      timeout: Duration.minutes(10),
    })
  }).toThrow('This lambda must timeout in 5 minutes or less');
});
