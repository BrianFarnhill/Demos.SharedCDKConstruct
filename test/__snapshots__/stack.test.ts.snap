// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches SnapShot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store.",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TestFunctionDemoFunction0712CEEE": Object {
      "DependsOn": Array [
        "TestFunctionDemoFunctionServiceRoleE448AD34",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "674142b286ba6bc74af152f69b339c76c9a00a5f58b5265e9ab180f4c22f66f0.zip",
        },
        "Description": "An example function that comes from a shared CDK construct",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestFunctionDemoFunctionServiceRoleE448AD34",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestFunctionDemoFunctionServiceRoleE448AD34": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
